### PR TITLE
Add Structured Logging

### LINKED ISSUE #3359: Stabilize structured logging
dbt can output structured JSON logs via:
- `dbt --log-output json` (CLI)
- `logs` parameter (RPC)

These logs include a lot of information _similar to_ dbt's JSON artifacts, with the crucial differences being that logs are written/returned in real time, whereas JSON artifacts are only written at the end of invocations.

As of v0.19, JSON artifacts are contracted and documented, but JSON logs are _not_ contracted and [barely documented](https://docs.getdbt.com/reference/global-cli-flags#log-formatting). We should change this ahead of v1.0. There are other technologies and community members who want to build on top of stable logs (e.g. dbt Cloud IDE, https://github.com/fishtown-analytics/dbt/issues/2915).

As I see it, the scope of this issue is to:
- Document structured (JSON) logs as they work today
- Communicate which components are stable (statuses, log-levels, node-level information) and which components are liable to change in future versions (exact language of info-level messages)
- Identify discrepancies between the information available in structure logs and the information included in JSON artifacts, and (if very slight) bring them into consistency

### PR DESCRIPTION
**Pull request by nathaniel-may**
resolves #3359

### Description

The `structured-logging` branch is a feature branch and small incremental changes should be merged via PR into this branch. Tests need to pass for each merge into this feature branch. There are many integration tests that rely on certain wording of log messages, and it is important that these tests remain intact by converting them to use codes incrementally.

#### Phase 1:

- [x] merge initial framework design
    - #4137
- [x] add adapter interface
    - ~#4161~ #4176
- [x] change every single callsite to the existing logger in the app and tests
    - #4164
    - #4163
    - #4165
    - #4166
    - #4169
    - #4183
    - #4199
    - #4219
- [x] plumb the cli color flag so structlog respects it
    - #4176
- [x] secret scrubbing
    - #4203
- [x] send all logs to stdout and dbt.log through some logger
    - ~#4176~ #4231
- [x] hook in the option for json logs
    - ~#4176~ #4231
- [x] update tests
    - #4257
- [ ] update adapter repos: (requires this PR to be merged first)
    - [redshift](https://github.com/dbt-labs/dbt-redshift/pull/34)
    - [snowflake](https://github.com/dbt-labs/dbt-snowflake/pull/42)
    - [bigquery](https://github.com/dbt-labs/dbt-bigquery/pull/55)
    - [spark](https://github.com/dbt-labs/dbt-spark/pull/251)

#### Phase 2:

- [ ] add additional log lines for node status reporting
- [ ] create scalable in-memory event history
- [ ] improve message wording
- [ ] fine tune `structlog` configs including look and feel
- [ ] add codes to each event
- [ ] add thread id to each event
- [ ] make sure we can distinguish logs from different runs (preserve process id like it is today, or allow some sort of user markers)
- [ ] upgrade rpc backdoor away from env vars ala #4055

### Checklist

- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [ ] I have run this code in development and it appears to resolve the stated issue
- [ ] This PR includes tests, or tests are not required/relevant for this PR
- [ ] I have updated the `CHANGELOG.md` and added information about my change

### CODE REVIEW

File: core/dbt/events/functions.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+global STDOUT_LOG
+STDOUT_LOG = logging.getLogger('default_stdout')
+STDOUT_LOG.setLevel(logging.INFO)
+stdout_handler = logging.StreamHandler()
+stdout_handler.setLevel(logging.INFO)
+STDOUT_LOG.addHandler(stdout_handler)
+global color
+format_color = True
+global json
+format_json = False
```
Comment (kwigley) [inline core/dbt/events/functions.py:31]: I don't understand why global is needed here. `global json` is referring to the json module here I think as well.

File: core/dbt/events/functions.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+
+def setup_event_logger(log_path):
```
Comment (kwigley) [inline core/dbt/events/functions.py:34]: I don't think `log_path` is being used in this function. Should it be used for `log_dest`?

File: core/dbt/events/functions.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
+
+from colorama import Style
+import dbt.events.functions as this  # don't worry I hate it too.
```
Comment (kwigley) [inline core/dbt/events/functions.py:3]: Why is this needed?

**Review by kwigley — COMMENTED** [review]
woo! left some questions for anyone who is familiar 😄

File: core/dbt/events/functions.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+
+def setup_event_logger(log_path):
+    make_log_dir_if_missing(log_path)
+    this.format_json = flags.LOG_FORMAT == 'json'
+    # USE_COLORS can be None if the app just started and the cli flags
+    # havent been applied yet
+    this.format_color = True if flags.USE_COLORS else False
+    # TODO this default should live somewhere better
+    log_dest = os.path.join('logs', 'dbt.log')
```
Comment (jtcohen6) [inline core/dbt/events/functions.py:41]: To Kyle's point above, shouldn't this be:
```suggestion
    log_dest = os.path.join(log_path, 'dbt.log')
```

File: core/dbt/events/functions.py (hunk 1 above)
Comment (nathaniel-may) [reply core/dbt/events/functions.py:31]: oh good catch. I was supposed to change that to global format_json, but it's working without that declaration. so I guess I don't need it.

File: core/dbt/events/functions.py (hunk 2 above)
Comment (nathaniel-may) [reply core/dbt/events/functions.py:34]: yup. it should.

File: core/dbt/events/functions.py (hunk 4 above)
Comment (nathaniel-may) [reply core/dbt/events/functions.py:41]: yup. this is correct.

File: core/dbt/events/functions.py (hunk 3 above)
Comment (nathaniel-may) [reply core/dbt/events/functions.py:3]: yeah I just tried deleting all the `this` calls, and flake8 gives me `F841 local variable 'format_json' is assigned to but never used` 🤷‍♂️

**Review by jtcohen6 — APPROVED** [review]
One failing test, `test_postgres_debug_with_profiles_separate_from_project_dir`, that only fails on Windows. It's definitely related to changes here, but I don't think it's worth holding off on this PR any longer. @nathaniel-may and I have a hunch that it's more likely a symptom of our testing scaffold than it is representative of an actual issue in dbt-core.

Let's open a new issue to investigate further, and add to the phase 2 to-do list. In the meantime, let's move ahead with this change.

Amazing work @nathaniel-may @iknox-fa @emmyoop, and to everyone who helped along the way :)
